File: Daily/129.sum-root-to-leaf-numbers.cpp
/*
 * @lc app=leetcode id=129 lang=cpp
 *
 * [129] Sum Root to Leaf Numbers
 */
struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};
// @lc code=start
class Solution {
 public:
  int sum;
  int currentNum;

  Solution() {
    sum = 0;
    currentNum = 0;
  }

  int sumNumbers(TreeNode* root) {
    backtrack(root);
    return sum;
  }

 private:
  void backtrack(TreeNode* root) {
    if (root == nullptr) {
      return;
    }

    currentNum *= 10;
    currentNum += root->val;
    if (root->left == nullptr && root->right == nullptr) {
      sum += currentNum;
    } else {
      backtrack(root->left);
      backtrack(root->right);
    }
    currentNum -= root->val;
    currentNum /= 10;
  }
};
// @lc code=end

File: src/2023-2024/129.sum-root-to-leaf-numbers.js
/*
 * @lc app=leetcode id=129 lang=javascript
 *
 * [129] Sum Root to Leaf Numbers
 */
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start

/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function (root) {
  let totalSum = 0;
  let currentPathSum = 0;

  /**
   * @param {TreeNode} root
   * @returns {void}
   */
  const backtrack = (root) => {
    if (root === null) return;

    currentPathSum *= 10;
    currentPathSum += root.val;
    if (root.left === null && root.right === null) {
      totalSum += currentPathSum;
    } else {
      backtrack(root.left);
      backtrack(root.right);
    }
    currentPathSum -= root.val;
    currentPathSum /= 10;
  };

  backtrack(root);
  return totalSum;
};
// @lc code=end
